apps/src/music/player/soundEffects.js

```diff
@@ -1,62 +1,165 @@
+    this.busses = {};
+    this.generateBusses();
   }
 
-  if (['low', 'medium'].includes(effects.filter)) {
-    const filter = audioContext.createBiquadFilter();
-    filter.type = 'lowpass';
-    filter.frequency.value = effects.filter === 'low' ? 800 : 3000;
-
-    // Connect the last node to this one.
-    lastNode.connect(filter);
-
-    // This is now the last node.
-    lastNode = filter;
+  // Generates a set of busses.  Because these effects can be
+  // expensive, we want to pre-generate the busses and then connect
+  // sounds to the appropriate one.
+  generateBusses() {
+    const busEffects = [
+      {filter: 'medium', delay: 'medium'},
+      {filter: 'low', delay: 'low'},
+      {filter: 'low', delay: 'medium'},
+      {filter: 'medium', delay: 'low'},
+      {filter: 'medium'},
+      {filter: 'low'},
+      {delay: 'medium'},
+      {delay: 'low'}
+    ];
+
+    busEffects.forEach(busEffect => {
+      let keyStrings = [];
+      if (busEffect.filter === 'medium') {
+        keyStrings.push('filter_medium');
+      }
+      if (busEffect.filter === 'low') {
+        keyStrings.push('filter_low');
+      }
+      if (busEffect.delay === 'medium') {
+        keyStrings.push('delay_medium');
+      }
+      if (busEffect.delay === 'low') {
+        keyStrings.push('delay_low');
+      }
+
+      const {firstNode, lastNode} = this.generateBus(busEffect);
+
+      // The bus connects to our main output.
+      lastNode.connect(this.audioContext.destination);
+
+      // Save the pre-generated bus.
+      const keyString = keyStrings.join('-');
+      this.busses[keyString] = firstNode;
+    });
   }
 
-  if (['low', 'medium'].includes(effects.delay)) {
-    // Create a node that does nothing so we can join the two outputs
-    // to it.
-    const output = audioContext.createGain();
-    output.gain.value = 1;
-
-    const delay = audioContext.createDelay();
-
-    // At 120 beats per minute, 0.25 seconds is half a beat.
-    delay.delayTime.value = 0.25;
-
-    const dry = audioContext.createGain();
-    const wet = audioContext.createGain();
-
-    const feedback = audioContext.createGain();
-    feedback.gain.value = effects.delay === 'low' ? 0.1 : 0.3;
-
-    lastNode.connect(dry);
-    dry.connect(output);
-
-    const filter = audioContext.createBiquadFilter();
-    filter.type = 'lowpass';
-    filter.frequency.value = 4000;
-
-    lastNode.connect(filter);
-    filter.connect(delay);
-    delay.connect(feedback);
-    feedback.connect(filter);
-
-    delay.connect(wet);
-    wet.connect(output);
-
-    // This is now the last node.
-    lastNode = output;
+  // Generate a bus for the given effects.
+  generateBus(effects) {
+    let firstNode = null;
+    let lastNode = null;
+
+    if (['low', 'medium'].includes(effects.filter)) {
+      const filter = this.audioContext.createBiquadFilter();
+      filter.type = 'lowpass';
+      filter.frequency.value = effects.filter === 'low' ? 800 : 3000;
+
+      if (lastNode) {
```

Comment:
Will `lastNode` ever be defined? Looks like it's always `null` at the start of this function?

Reply:
It won't.  I had done this deliberately to create a kind of duplicatable template, but it really is unreachable code, so I'll remove it.